### PR TITLE
Admins can view all users

### LINKED ISSUE #5: Admin can view all users
List of users. Default filtered users with login credentials.

### PR DESCRIPTION
**Pull request by MJ1400**
Adminid saavad näha andmebaasis olevate kasutajate ees- ja perekonnanimesid.

closes #5

### CODE REVIEW

**Comment by romilrobtsenkov** [comment]
Please fix linting before the review, otherwise, too many lines changed. Use eslint plugin only for linting.
Screenshot 2020-12-16 at 18 39 12

File: teemaderegister-be/src/controllers/users.js [hunk 1]
```diff
@@ -24,15 +24,15 @@ module.exports.getUser = async (req, res) => {
         slug: user.profile.slug,
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/users.js:24]: many unnecessary changes in this file still

File: teemaderegister-be/src/routes.js [hunk 2]
```diff
@@ -16,33 +16,98 @@ const topics = require('./controllers/topics')
 const users = require('./controllers/users')
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/routes.js:16]: many unnecessary changes in this file still

File: teemaderegister-be/src/routes.js [hunk 3]
```diff
@@ -16,33 +16,98 @@ const topics = require('./controllers/topics')
...
 // SAMPLE
-router.get('/admin/', jwtEnsure, allowRoles([ADMIN]), asyncMiddleware(admin.getSecret))
+router.get(
+  '/admin/',
+  jwtEnsure,
+  allowRoles([ADMIN]),
+  asyncMiddleware(admin.getSecret)
+)
+
+router.get(
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/routes.js:106]: I would use route `/admin/users`

File: teemaderegister-be/src/controllers/users.js [hunk 4]
```diff
@@ -174,8 +171,19 @@ module.exports.resetPicture = async (req, res) => {
...
 }
+
+module.exports.getAllUsers = async (req, res) => {
+  const users = await User.find({})
+
+  let data = []
+  for (const number in users) {
+    data.push(users[number]['profile'])
+  }
+  return res.json(data)
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/users.js:188]: I would return data format as a user object as it is returned from DB, still including a unique identifier. Modifying user object too much is not a too good idea.

File: teemaderegister-be/src/controllers/users.js [hunk 5]
```diff
@@ -174,8 +171,19 @@ module.exports.resetPicture = async (req, res) => {
...
-    log.debug(`${err && err.code} - Probably ${original} already deleted`))
+  await fs.unlink(`${process.env.UPLOAD_DIR + original}`, (err) =>
+    log.debug(`${err && err.code} - Probably ${original} already deleted`)
+  )
 
   return res.json({ user, message: 'Picture updated successfully' })
 }
+
+module.exports.getAllUsers = async (req, res) => {
+  const users = await User.find({})
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/users.js:182]: Here you can select or deselect the fields you do not need. I would exclude password. Others may be valid for admin to see.

File: teemaderegister-react/src/components/Admin.jsx [hunk 6]
```diff
@@ -16,20 +16,20 @@ const propTypes = {
   auth: shape({
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/Admin.jsx:16]: many unnecessary changes in this file still

File: teemaderegister-react/src/components/ViewAllUsers.jsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    key: 'lastName',
+  },
+]
+
+function ViewAllUsers() {
+  const [users, setUsers] = useState([])
+
+  useEffect(() => {
+    const getUserData = async () => {
+      var jwt = localStorage.getItem('jwtToken')
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/ViewAllUsers.jsx:22]: Instead of using this, there is an API utils (/util/Api.js) helper that does those things (url + jwt) for you.

Otherwise, You can use this approach, or I would recommend using REDUX as Home pages do. You can keep this component functional and trigger request on useEffect.

Please add loading!

Comment (MJ1400) [reply teemaderegister-react/src/components/ViewAllUsers.jsx:22]: There's no such file as Api.js? And what exactly do you mean by "add loading"?

utils

Comment (MJ1400) [reply teemaderegister-react/src/components/ViewAllUsers.jsx:22]: I don't quite understand how to use the helper. And what do you mean by "add loading"?

Comment (raimop) [reply teemaderegister-react/src/components/ViewAllUsers.jsx:22]: [CurriculumAdd.jsx](https://github.com/rakenduste-programmeerimine-2020s/teemaderegister/blob/main/teemaderegister-react/src/components/CurriculumAdd.jsx)

There is an example of using the API utils in the "fetchUsers" function

"Add loading" => I'm presuming that you'd have a loading state that is initially false and while it starts fetching the users the loading state is true and will be turned to false again once it's all done. Preferably also showing a spinner while the loading is true.

File: teemaderegister-be/src/controllers/users.js [hunk 8]
```diff
@@ -179,3 +179,9 @@ module.exports.resetPicture = async (req, res) => {
...
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/users.js:184]: You should never return user passwords.

Comment (romilrobtsenkov) [reply teemaderegister-be/src/controllers/users.js:184]: it is possible to exclude them with 
```
select('-field')
```

File: teemaderegister-react/src/components/Admin.jsx [hunk 9]
```diff
@@ -102,11 +102,11 @@ class Admin extends React.Component {
                 </Menu.Item>
                 {(isAdmin || isStudyAssistant) &&
                 <SubMenu key='users' title={<span><UserOutlined/>Users</span>}>
+                  <Menu.Item key='allusers'>All users</Menu.Item>
                   <Menu.Item
                     key='addNewUser'
                     icon={<UserAddOutlined/>}>Add new user</Menu.Item>
-                  <Menu.Item key='supervisors'>Supervisors</Menu.Item>
-                  <Menu.Item key='students'>Students</Menu.Item>
+
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/Admin.jsx:109]: unnessecery empty row

**Review by romilrobtsenkov — COMMENTED** [review]
As of today the view did not load
Screenshot 2021-01-03 at 15 16 41

File: teemaderegister-react/src/components/ViewAllUsers.jsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+]
+
+export default function ViewAllUsers () {
+  const [users, setUsers] = useState([])
+  const [isLoading, setIsLoading] = useState(true)
+
+  useEffect(() => {
+    API('GET', ADMIN_VIEW_USERS_URL)
+      .then(res => {
+        const userInfo = res.map((userData) => ({
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/ViewAllUsers.jsx:26]: I would pass userdata directly to setUsers

File: teemaderegister-react/src/components/ViewAllUsers.jsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+          key: userData._id,
+          firstName: userData.profile.firstName,
+          lastName: userData.profile.lastName
+        }))
+        setUsers(userInfo)
+        setIsLoading(false)
+      })
+  }, [])
+
+  return <Table dataSource={users} columns={columns} loading={isLoading}/>
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/ViewAllUsers.jsx:36]: Here You can add rowkey via `rowKey={r => r._id}`

Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/ViewAllUsers.jsx:36]: Here You can add rowkey via `rowKey={r => r._id}`

File: teemaderegister-react/src/components/ViewAllUsers.jsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  {
+    title: 'First Name',
+    dataIndex: 'firstName',
+    key: 'firstName'
+  },
+  {
+    title: 'Last Name',
+    dataIndex: 'lastName',
+    key: 'lastName'
+  }
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/ViewAllUsers.jsx:16]: You can use dataIndex only if it is unique, no need to use key then

File: teemaderegister-be/src/controllers/users.js [hunk 13]
```diff
@@ -179,3 +179,9 @@ module.exports.resetPicture = async (req, res) => {
 
   return res.json({ user, message: 'Picture updated successfully' })
 }
+
+module.exports.getAllUsers = async (req, res) => {
+  const users = await User.find({}, {login: 0})
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/users.js:184]: As the page is for admin only, I would exclude only password and include other columns, email if there is any etc.

As the could be many users would do also pagination – no need to add it now, can be done with separate task later.

**Review by romilrobtsenkov — APPROVED** [review]
Otherwise looks good. Please do those minor changes and merge to master.

File: teemaderegister-be/src/controllers/users.js (hunk 13 above)
Comment (MJ1400) [reply teemaderegister-be/src/controllers/users.js]: What do you exactly mean by pagination? There is already pagination already included in the ant.design table or am I misunderstanding something?
 
pagination

File: teemaderegister-react/src/components/ViewAllUsers.jsx (hunk 10 above)
Comment (MJ1400) [reply teemaderegister-react/src/components/ViewAllUsers.jsx:26]: Without mapping the response or how?
